File: ios/Voxa/ViewModels/ChatViewModel.swift
import Foundation
import SwiftUI

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [VoxaMessage] = []
    @Published var isLoading = false
    @Published var isSending = false
    @Published var error: String?
    @Published var typingUsers: [String] = []  // usernames currently typing

    private var currentChannelId: String?
    private var pollTask: Task<Void, Never>?
    private let storageKey = "voxa_msgs_v2_"
    private let socket = SocketClient.shared

    func load(channelId: String, token: String?) async {
        currentChannelId = channelId
        isLoading = true
        messages = loadLocal(channelId: channelId)

        do {
            let remote = try await APIClient.shared.messages(channelId: channelId)
            messages = remote
            saveLocal(channelId: channelId)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false

        // Wire up real-time socket callbacks
        socket.onNewMessage = { [weak self] msg in
            guard let self, msg.channelId == channelId else { return }
            self.handleNewMessage(msg)
        }
        socket.onMessageEdit = { [weak self] msg in
            guard let self, msg.channelId == channelId else { return }
            if let idx = self.messages.firstIndex(where: { $0.id == msg.id }) {
                self.messages[idx] = msg
                self.saveLocal(channelId: channelId)
            }
        }
        socket.onMessageDelete = { [weak self] id in
            guard let self else { return }
            self.messages.removeAll { $0.id == id }
            self.saveLocal(channelId: channelId)
        }
        socket.onTypingUpdate = { [weak self] cid, _, username, typing in
            guard let self, cid == channelId else { return }
            if typing {
                if !self.typingUsers.contains(username) {
                    self.typingUsers.append(username)
                }
            } else {
                self.typingUsers.removeAll { $0 == username }
            }
        }

        // Re-join the channel after reconnect
        socket.onConnect = { [weak self] in
            guard let self else { return }
            Task { @MainActor in self.socket.joinChannel(channelId) }
        }
        socket.joinChannel(channelId)

        // Fallback poll every 10s in case socket misses anything
        startPolling(channelId: channelId)
    }

    func send(content: String, channelId: String, author: User) async {
        guard !content.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        isSending = true

        let optimistic = VoxaMessage(
            id: "opt_\(UUID().uuidString)",
            content: content,
            authorId: author.id,
            author: author.username,
            displayName: author.displayName,
            avatarUrl: author.avatarUrl,
            avatarColor: author.avatarColor,
            discriminator: author.discriminator,
            channelId: channelId,
            timestamp: Date(),
            edited: false, editedAt: nil,
            parentId: nil, replyCount: 0, reactions: nil
        )
        messages.append(optimistic)

        do {
            let real = try await APIClient.shared.sendMessage(channelId: channelId, content: content)
            if let idx = messages.firstIndex(where: { $0.id == optimistic.id }) {
                messages[idx] = real
            }
            saveLocal(channelId: channelId)
        } catch {
            // Keep optimistic message visible
        }
        isSending = false
    }

    func edit(message: VoxaMessage, newContent: String) async {
        guard let idx = messages.firstIndex(where: { $0.id == message.id }) else { return }
        messages[idx].content = newContent
        messages[idx].edited = true
        if let cid = currentChannelId { saveLocal(channelId: cid) }

        do {
            let updated = try await APIClient.shared.editMessage(messageId: message.id, content: newContent)
            if let i = messages.firstIndex(where: { $0.id == message.id }) {
                messages[i] = updated
            }
        } catch {}
    }

    func delete(message: VoxaMessage) async {
        messages.removeAll { $0.id == message.id }
        if let cid = currentChannelId { saveLocal(channelId: cid) }
        try? await APIClient.shared.deleteMessage(messageId: message.id)
    }

    func disconnect() {
        if let cid = currentChannelId {
            socket.leaveChannel(cid)
        }
        pollTask?.cancel()
        pollTask = nil
        socket.onNewMessage = nil
        socket.onMessageEdit = nil
        socket.onConnect = nil
        socket.onMessageDelete = nil
        socket.onTypingUpdate = nil
        currentChannelId = nil
        typingUsers = []
    }

    // MARK: - Private helpers

    private func handleNewMessage(_ msg: VoxaMessage) {
        // Replace matching optimistic message or append
        if let optIdx = messages.firstIndex(where: {
            $0.id.hasPrefix("opt_") && $0.content == msg.content && $0.authorId == msg.authorId
        }) {
            messages[optIdx] = msg
        } else if !messages.contains(where: { $0.id == msg.id }) {
            messages.append(msg)
        }
        if let cid = currentChannelId { saveLocal(channelId: cid) }
    }

    private func startPolling(channelId: String) {
        pollTask?.cancel()
        pollTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 10_000_000_000) // 10s fallback
                guard !Task.isCancelled else { break }
                do {
                    let fresh = try await APIClient.shared.messages(channelId: channelId, limit: 50)
                    if !Task.isCancelled {
                        mergeMessages(fresh, channelId: channelId)
                    }
                } catch {}
            }
        }
    }

    private func mergeMessages(_ fresh: [VoxaMessage], channelId: String) {
        var merged = messages
        for msg in fresh {
            if let idx = merged.firstIndex(where: { $0.id == msg.id }) {
                merged[idx] = msg
            } else if !merged.contains(where: { $0.id == msg.id }) {
                merged.append(msg)
            }
        }
        merged.removeAll { msg in
            msg.id.hasPrefix("opt_") &&
            fresh.contains(where: { $0.content == msg.content && $0.authorId == msg.authorId })
        }
        messages = merged.sorted { $0.timestamp < $1.timestamp }
        saveLocal(channelId: channelId)
    }

    // MARK: - Local Persistence

    private func saveLocal(channelId: String) {
        let toSave = Array(messages.suffix(200))
        if let data = try? JSONEncoder().encode(toSave) {
            UserDefaults.standard.set(data, forKey: storageKey + channelId)
        }
    }

    private func loadLocal(channelId: String) -> [VoxaMessage] {
        guard let data = UserDefaults.standard.data(forKey: storageKey + channelId),
              let msgs = try? JSONDecoder().decode([VoxaMessage].self, from: data) else { return [] }
        return msgs
    }
}

File: android/app/src/main/java/com/voxa/app/viewmodel/ChatViewModel.kt
package com.voxa.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.voxa.app.data.api.ApiClient
import com.voxa.app.data.api.SocketClient
import com.voxa.app.data.model.User
import com.voxa.app.data.model.VoxaMessage
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class ChatViewModel : ViewModel() {

    private val _messages = MutableStateFlow<List<VoxaMessage>>(emptyList())
    val messages: StateFlow<List<VoxaMessage>> = _messages

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _isSending = MutableStateFlow(false)
    val isSending: StateFlow<Boolean> = _isSending

    private val _typingUsers = MutableStateFlow<List<String>>(emptyList())
    val typingUsers: StateFlow<List<String>> = _typingUsers

    private var currentChannelId: String? = null
    private var pollJob: Job? = null

    fun load(channelId: String) {
        currentChannelId = channelId
        setupSocket(channelId)
        viewModelScope.launch {
            _isLoading.value = true
            ApiClient.messages(channelId)
                .onSuccess { _messages.value = it }
            _isLoading.value = false
        }
        startPolling(channelId)
    }

    fun send(content: String, channelId: String, author: User) {
        if (content.isBlank()) return
        viewModelScope.launch {
            _isSending.value = true
            val optimistic = VoxaMessage(
                id = "opt_${System.currentTimeMillis()}",
                content = content,
                authorId = author.id,
                author = author.username,
                displayName = author.displayName,
                avatarColor = author.avatarColor,
                channelId = channelId,
                timestamp = java.time.Instant.now().toString(),
                edited = false,
            )
            _messages.value = _messages.value + optimistic

            ApiClient.sendMessage(channelId, content)
                .onSuccess { real ->
                    _messages.value = _messages.value.map {
                        if (it.id == optimistic.id) real else it
                    }
                }
            _isSending.value = false
        }
    }

    fun edit(message: VoxaMessage, newContent: String) {
        viewModelScope.launch {
            _messages.value = _messages.value.map {
                if (it.id == message.id) it.copy(content = newContent, edited = true) else it
            }
            ApiClient.editMessage(message.id, newContent)
                .onSuccess { updated ->
                    _messages.value = _messages.value.map {
                        if (it.id == message.id) updated else it
                    }
                }
        }
    }

    fun delete(message: VoxaMessage) {
        _messages.value = _messages.value.filter { it.id != message.id }
        viewModelScope.launch { ApiClient.deleteMessage(message.id) }
    }

    fun disconnect() {
        currentChannelId?.let { SocketClient.leaveChannel(it) }
        pollJob?.cancel()
        pollJob = null
        currentChannelId = null
        SocketClient.onNewMessage = null
        SocketClient.onMessageEdit = null
        SocketClient.onMessageDelete = null
        SocketClient.onTypingUpdate = null
        _messages.value = emptyList()
        _typingUsers.value = emptyList()
    }

    private fun setupSocket(channelId: String) {
        SocketClient.joinChannel(channelId)

        SocketClient.onNewMessage = { msg ->
            if (msg.channelId == channelId) {
                val existing = _messages.value
                val optIdx = existing.indexOfFirst {
                    it.id.startsWith("opt_") && it.content == msg.content && it.authorId == msg.authorId
                }
                _messages.value = if (optIdx >= 0) {
                    existing.toMutableList().apply { set(optIdx, msg) }
                } else if (existing.none { it.id == msg.id }) {
                    existing + msg
                } else existing
            }
        }
        SocketClient.onMessageEdit = { msg ->
            if (msg.channelId == channelId) {
                _messages.value = _messages.value.map { if (it.id == msg.id) msg else it }
            }
        }
        SocketClient.onMessageDelete = { id ->
            _messages.value = _messages.value.filter { it.id != id }
        }
        SocketClient.onTypingUpdate = { cid, _, username, typing ->
            if (cid == channelId) {
                _typingUsers.value = if (typing)
                    (_typingUsers.value + username).distinct()
                else
                    _typingUsers.value - username
            }
        }
    }

    private fun startPolling(channelId: String) {
        pollJob?.cancel()
        pollJob = viewModelScope.launch {
            while (isActive) {
                delay(10_000)
                ApiClient.messages(channelId)
                    .onSuccess { fresh ->
                        val current = _messages.value
                        val merged = (current + fresh).distinctBy { it.id }
                            .sortedBy { it.timestamp }
                            .filter { msg ->
                                if (!msg.id.startsWith("opt_")) return@filter true
                                fresh.none { it.content == msg.content && it.authorId == msg.authorId }
                            }
                        _messages.value = merged
                    }
            }
        }
    }
}
